File: format/index_description.go
package format

import (
	"database/sql"
	"fmt"
)

// IndexDescription contains all the data known about a specific index.
// Note that some indexes may be related (e.g. in cases of clustered indexes).
type IndexDescription struct {
	Table      string         `db:"Table"`
	NonUnique  bool           `db:"Non_unique"`
	KeyName    string         `db:"Key_name"`
	SeqInIndex int            `db:"Seq_in_index"`
	ColumnName string         `db:"Column_name"`
	Comment    sql.NullString `db:"Comment"`

	// Not used (yet)
	Collation    sql.NullString `db:"Collation"`
	Cardinality  sql.NullString `db:"Cardinality"`
	SubPart      sql.NullString `db:"Sub_part"`
	Packed       sql.NullString `db:"Packed"`
	Null         sql.NullString `db:"Null"`
	IndexType    sql.NullString `db:"Index_type"`
	IndexComment sql.NullString `db:"Index_comment"`
	Visible      sql.NullString `db:"Visible"`
	Expression   sql.NullString `db:"Expression"`
}

// IndexDescriptions is a set of index descriptions
type IndexDescriptions []IndexDescription

// ConvertToLogicalIndexes converts raw index descriptions into a aggregated
// type.
func (descs IndexDescriptions) ConvertToLogicalIndexes() ([]LogicalIndex, error) {
	indices := map[string]LogicalIndex{}

	for _, description := range descs {
		var li LogicalIndex
		var ok bool
		if li, ok = indices[description.KeyName]; !ok {
			li = LogicalIndex{
				Table:     description.Table,
				NonUnique: description.NonUnique,
				KeyName:   description.KeyName,
				Comment:   description.Comment.String,
			}
		}

		li.IndexedColumnNamesOrdered = append(li.IndexedColumnNamesOrdered, description.ColumnName)

		if len(li.IndexedColumnNamesOrdered) != description.SeqInIndex {
			return nil, fmt.Errorf("internal logic error: expecting for indexed columns to always be returned in sequence")
		}

		indices[description.KeyName] = li
	}

	// return things in the order they were received -- required for the out
	// put to be deterministic.
	result := []LogicalIndex{}
	var addedKeys = map[string]struct{}{}
	for _, description := range descs {

		// only add keys once
		if _, ok := addedKeys[description.KeyName]; !ok {
			result = append(result, indices[description.KeyName])
			addedKeys[description.KeyName] = struct{}{}
		}
	}

	return result, nil
}

// LogicalIndex defines a "logical index -- e.g. what will make the most sense
// to humans"
type LogicalIndex struct {
	Table                     string
	NonUnique                 bool
	KeyName                   string
	IndexedColumnNamesOrdered []string
	Comment                   string
}

File: src/format/index_description.rs
use anyhow::Result;

#[derive(Debug, Clone, PartialEq)]
pub struct IndexDescription {
    pub table: String,
    pub non_unique: bool,
    pub key_name: String,
    pub seq_in_index: i32,
    pub column_name: Option<String>,
    pub comment: Option<String>,
    pub expression: Option<String>,
}

#[derive(Debug)]
pub struct IndexDescriptions(pub Vec<IndexDescription>);

#[derive(Debug, Clone, PartialEq)]
pub struct LogicalIndex {
    pub table: String,
    pub non_unique: bool,
    pub key_name: String,
    pub indexed_column_names_ordered: Vec<String>,
    pub comment: String,
    pub expression: String,
}

impl IndexDescriptions {
    /// Convert raw index descriptions into aggregated type
    pub fn convert_to_logical_indexes(&self) -> Result<Vec<LogicalIndex>> {
        let mut indices = std::collections::HashMap::new();

        for description in &self.0 {
            let li = indices.entry(&description.key_name).or_insert_with(|| LogicalIndex {
                table: description.table.clone(),
                non_unique: description.non_unique,
                key_name: description.key_name.clone(),
                indexed_column_names_ordered: Vec::new(),
                comment: description.comment.clone().unwrap_or_default(),
                expression: description.expression.clone().unwrap_or_default(),
            });

            if let Some(column_name) = &description.column_name {
                // Check if the number of indexed columns equals the sequence number
                if li.indexed_column_names_ordered.len() + 1 != description.seq_in_index as usize {
                    return Err(anyhow::anyhow!(
                        "internal logic error: expecting for indexed columns to always be returned in sequence"
                    ));
                }
                li.indexed_column_names_ordered.push(column_name.clone());
            }
        }

        // Return things in the order they were received for deterministic output
        let mut result = Vec::new();
        let mut added_keys = std::collections::HashSet::new();
        
        for description in &self.0 {
            if !added_keys.contains(&description.key_name) {
                if let Some(li) = indices.get(&description.key_name) {
                    result.push(li.clone());
                    added_keys.insert(description.key_name.clone());
                }
            }
        }

        Ok(result)
    }
}


